### ISSUE TITLE
Tuned .cact decodes differently in the engine than in JAX with identical tokens, weights and first-token logits (trained refusals become tool calls)

### ISSUE THREAD

**Issue by alexnodeland** [cactus-compute/needle#117]
## Summary

A QAT LoRA adapter trained with `cactus-needle==2.0.13` (`--qat-bits auto`, the post-#91 template) and exported with `needle build` behaves differently in the native engine than under greedy decoding of the same weights in JAX, on its own training rows. Tool-call rows agree exactly, arguments included. Refusal rows do not: JAX reproduces the trained refusal (`<tool_call>[]</tool_call>`), the engine emits a tool call with a plausible fine-tuned-style reasoning on 7 of 8 rows. The engine is deterministic.

I have checked the usual suspects and they match, so this is a counterexample to the way #91 was closed (template + numerics):

- **Tokens.** `needle_init` returns the same prefix token count as the training template for every variant I tried (with/without system, compact vs spaced tools, different system lengths). With `NEEDLE_DEBUG=1`, the prefix ids and turn ids the engine prints are identical to `[BOS] + tokenizer.encode(render_example(...)[0])`, 393 of 393 tokens.
- **Weights.** `needle.model.export.read_export(model.cact)` dequantised equals `cq_ste_mixed_params(merge_lora(params, lora, scale))` element for element when the merge is done in the checkpoint's float16, as `build_main` does. (Aside: training merges in float32, so the QAT forward saw slightly different 2-bit indices — about 6% of them — than the export ships. Not the cause here: JAX decoding of the float16-merged weights also refuses.)
- **First-token logits.** The `NEEDLE_DEBUG` dump of the first generated position matches the JAX forward (`quant=True`): mean |Δ| 0.04 for the base weights, 0.54 for the tuned weights, same top-5, engine ≈ 1.017 × JAX + 0.07. So the prefill is the same computation.
- **Not the window, cache or threads.** `NEEDLE_KV_WINDOW` 128/512/2048, `NEEDLE_KV_BITS` 16/32, `NEEDLE_THREADS=1`, `NEEDLE_NO_REBASE=1`, and `needle_load`/`needle_init` call order all leave the output unchanged.
- **Not the prefix cache.** After a first completion I sent a tool-result turn and compared the dump of that second call with JAX states built from a tuned prefix vs a base-weight prefix; the tuned prefix matches (mean |Δ| 1.5 vs 5–9), so the cache the engine decodes against is the tuned one.

What I cannot see is the logits after the first generated token: in JAX the refusal decision is one token (`gen…` at 94% vs `'` at 4% after `<think>\n`) and the engine picks `'` every time. Either the per-token decode path computes something different from prefill, or the decode policy is not plain argmax. A `NEEDLE_DEBUG` dump of every step's logits (not only the first) would settle this quickly.

## Environment

macOS 15 / Apple M3 Max, `cactus-needle` 2.0.13, engine 2.0.4 (`~/.cache/cactus-needle/v2/2.0.4/libneedle.dylib`, macos-arm64), JAX 0.4.38 on CPU, checkpoint `Cactus-Compute/needle2` `checkpoints/needle2.pkl` (`weight_bits=embedding=4,mhc=4,default=2`, `kv_window=256`).

## Reproduction

Everything is public on the Hub: `alexnodeland/site-needle` (`site-needle.cact`, the `adapter.pkl` it was built from, `tools.json`, `system.txt`). The script decodes eight training rows both ways.

```bash
pip install "cactus-needle[train]==2.0.13" huggingface_hub
JAX_PLATFORMS=cpu python repro.py
```

```python
"""Engine vs JAX greedy decode for a QAT LoRA adapter, on its own training rows.
pip install "cactus-needle[train]==2.0.13" huggingface_hub ; JAX_PLATFORMS=cpu python repro.py"""
import json, pickle, numpy as np, jax, jax.numpy as jnp, needle
from huggingface_hub import hf_hub_download
from needle.model.run import load_checkpoint
from needle.model.finetune import merge_lora, render_example
from needle.model.quantize import cq_ste_mixed_params, parse_bits_map, configure_deploy
from needle.model.architecture import SimpleAttentionNetwork
from needle.model.tokenizer import get_tokenizer, BOS_ID, EOS_ID, IM_END

REPO, REV = "alexnodeland/site-needle", "main"
cact = hf_hub_download(REPO, "site-needle.cact", revision=REV)
adapter = pickle.load(open(hf_hub_download(REPO, "adapter.pkl", revision=REV), "rb"))
tools = json.load(open(hf_hub_download(REPO, "tools.json", revision=REV)))
system = open(hf_hub_download(REPO, "system.txt", revision=REV)).read().strip()
ckpt = hf_hub_download("Cactus-Compute/needle2", "checkpoints/needle2.pkl")

params, config = load_checkpoint(ckpt)            # float16 params, as `needle build` merges them
lora = {tuple(k.split("/")): {"A": jnp.asarray(v["A"]), "B": jnp.asarray(v["B"])} for k, v in adapter["lora"].items()}
merged = merge_lora(params, lora, adapter["scale"])
bits_map, default = parse_bits_map(adapter["qat_bits_map"])
configure_deploy(act_bits=config.act_bits, kv_bits=config.kv_bits)
config.dtype = "float32"
q = cq_ste_mixed_params(jax.tree.map(lambda a: jnp.asarray(np.asarray(a, np.float32)), merged), bits_map, default)
model, tok = SimpleAttentionNetwork(config), get_tokenizer(config.vocab_size)
im_end = tok.encode(IM_END)

@jax.jit
def logits_at(ids):  # fixed length so the loop compiles once
    return model.apply({"params": q}, ids[None], quant=True)[0]

def jax_greedy(query, n=48):
    prompt, _ = render_example({"query": query, "tools": tools, "system": system, "answers": []})
    ids = [BOS_ID] + tok.encode(prompt)
    out = []
    for _ in range(n):
        padded = jnp.asarray(ids + out + [0] * (512 - len(ids) - len(out)), jnp.int32)
        nxt = int(jnp.argmax(logits_at(padded)[len(ids) + len(out) - 1])); out.append(nxt)
        if nxt == EOS_ID or out[-len(im_end):] == im_end: break
    return tok.decode(out).replace("\n", " ")

engine = needle.Needle(tools=tools, system=system, weights=cact)
refusals = ["what's the capital of France?", "who won the world cup in 2022?", "what time is it in Singapore?",
            "how do I bake sourdough?", "give me a recipe for banana bread", "what's 17 times 23?"]
calls = ["did Alex work at perch insights?", "what was Alex doing at Perch Insights?"]
for query in refusals + calls:
    e = engine.complete(query); engine.reset()
    print(f"\nQ: {query}\n  engine: calls={json.dumps(e['function_calls'])}  reasoning={e['reasoning']!r}\n  jax   : {jax_greedy(query)}")
```

Output on my machine:

```
Q: what's the capital of France?
  engine: calls=[{"name": "search_site", "arguments": {"section": "skills"}}]  reasoning="'France' -> topic; 'capital of France' -> section writing"
  jax   : <think> general knowledge; no site tool answers it </think> <tool_call>[]</tool_call><|im_end|>

Q: what time is it in Singapore?
  engine: calls=[{"name": "lookup_role", "arguments": {"company": "Singapore"}}]  reasoning="'Singapore' -> company"
  jax   : <think> general knowledge; no site tool answers it </think> <tool_call>[]</tool_call><|im_end|>

Q: how do I bake sourdough?
  engine: calls=[{"name": "search_site", "arguments": {"section": "skills"}}]  reasoning="'sourdough' -> section consulting"
  jax   : <think> general knowledge; no site tool answers it </think> <tool_call>[]</tool_call><|im_end|>

Q: did Alex work at perch insights?
  engine: calls=[{"name": "lookup_role", "arguments": {"company": "perch insights"}}]  reasoning="'perch insights' -> company"
  jax   : <think> 'perch insights' -> company </think> <tool_call>[{"name":"lookup_role","arguments":{"company":"perch insights"}}]</tool_call><|im_end|>
```

(The remaining rows follow the same pattern: six refusals become calls, both tool rows agree exactly. The `section: "skills"` in the engine's calls is its enum grounding overriding the model's `writing`/`consulting`; the divergence is already in the reasoning text.)

Happy to run anything else on the same artefacts.
